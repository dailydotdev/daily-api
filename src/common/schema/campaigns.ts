import z from 'zod';

export const CAMPAIGN_VALIDATION_SCHEMA = z.object({
  budget: z
    .int()
    .min(1000)
    .max(100000)
    .refine((value) => value % 1000 === 0, {
      error: 'Budget must be divisible by 1000',
    }),
  duration: z.int().min(1).max(30),
});
