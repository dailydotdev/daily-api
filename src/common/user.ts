import { DataSource } from 'typeorm';
import {
  Alerts,
  ArticlePost,
  Bookmark,
  BookmarkList,
  Comment,
  DevCard,
  Feed,
  Post,
  PostReport,
  Settings,
  SourceDisplay,
  SourceRequest,
  User,
  View,
} from '../entity';
import { ghostUser } from './index';
import { cancelSubscription } from './paddle';
import type { AuthContext, Context } from '../Context';
import { ForbiddenError } from 'apollo-server-errors';
import { logger } from '../logger';
import { CoresRole } from '../types';
import { remoteConfig } from '../remoteConfig';
import { UserTransaction } from '../entity/user/UserTransaction';
import { SubscriptionProvider, SubscriptionStatus } from './plus';
import { deleteUserResume } from './googleCloud';

export const deleteUser = async (
  con: DataSource,
  userId: string,
  messageId?: string,
) => {
  try {
    const { subscriptionFlags } = await con.getRepository(User).findOneOrFail({
      select: ['subscriptionFlags'],
      where: { id: userId },
    });

    // If the user has a subscription, we need to handle it before deletion
    if (subscriptionFlags?.subscriptionId) {
      if (
        subscriptionFlags?.provider === SubscriptionProvider.AppleStoreKit &&
        subscriptionFlags?.status === SubscriptionStatus.Active
      ) {
        throw new Error(
          'Apple subscriptions are not supported for user deletion',
        );
      }

      if (subscriptionFlags?.provider === SubscriptionProvider.Paddle) {
        const isGifted = !!subscriptionFlags.giftExpirationDate;
        // gifted subscription is a one-time payment hence not considered subscription in Paddle's terms
        if (!isGifted) {
          await cancelSubscription({
            subscriptionId: subscriptionFlags.subscriptionId,
          });
        }
        logger.info(
          {
            provider: SubscriptionProvider.Paddle,
            userId,
            isGifted,
            subscriptionId: subscriptionFlags.subscriptionId,
          },
          'Subscription cancelled user deletion',
        );
      }
    }

    // Delete user's resume if exists
    await deleteUserResume(userId);

    await con.transaction(async (entityManager): Promise<void> => {
      await entityManager.getRepository(View).delete({ userId });
      await entityManager.getRepository(Alerts).delete({ userId });
      await entityManager.getRepository(BookmarkList).delete({ userId });
      await entityManager.getRepository(Bookmark).delete({ userId });
      await entityManager.getRepository(Comment).update(
        { userId },
        {
          userId: ghostUser.id,
        },
      );
      await entityManager.getRepository(Comment).delete({ userId });
      await entityManager.getRepository(DevCard).delete({ userId });
      await entityManager.getRepository(Feed).delete({ userId });
      await entityManager.getRepository(PostReport).delete({ userId });
      await entityManager.getRepository(Settings).delete({ userId });
      await entityManager.getRepository(SourceDisplay).delete({ userId });
      await entityManager.getRepository(SourceRequest).delete({ userId });
      await entityManager
        .getRepository(ArticlePost)
        .update({ authorId: userId }, { authorId: null });
      // Manually set shared post to 404 dummy user
      await entityManager
        .getRepository(Post)
        .update({ authorId: userId }, { authorId: ghostUser.id });
      await entityManager
        .getRepository(Post)
        .update({ scoutId: userId }, { scoutId: null });
      await entityManager.getRepository(UserTransaction).update(
        {
          senderId: userId,
        },
        {
          senderId: ghostUser.id,
        },
      );
      await entityManager.getRepository(UserTransaction).update(
        {
          receiverId: userId,
        },
        {
          receiverId: ghostUser.id,
        },
      );
      await entityManager.getRepository(User).delete(userId);
    });
    logger.info(
      {
        userId,
        messageId,
      },
      'deleted user',
    );
  } catch (err) {
    logger.error(
      {
        userId,
        messageId,
        err,
      },
      'failed to delete user',
    );
    throw err;
  }
};

export /**
 * Function creator that wraps a function with auth protection
 * In this case it checks ctx contains userId but it can be expanded
 * in the future to check more advanced things.
 * This is a DX layer to make sure auth required stuff like cores
 * is protected on the function level.
 *
 * @template Props
 * @template Result
 * @param {(props: Props) => Result} fn
 * @return {Result}
 */
const createAuthProtectedFn = <
  Props extends { ctx: Pick<AuthContext, 'userId'> },
  Result,
>(
  fn: (props: Props) => Result,
) => {
  return (props: Props): Result => {
    if (!props.ctx.userId) {
      throw new ForbiddenError('Auth is required');
    }

    return fn(props);
  };
};

export const getUserCoresRole = ({ region }: { region?: string }) => {
  if (!region) {
    return CoresRole.None;
  }

  const ruleForRegion = remoteConfig.vars.coresRoleRules?.find((rule) =>
    rule.regions.includes(region),
  );

  if (ruleForRegion) {
    return ruleForRegion.role;
  }

  return CoresRole.Creator;
};

export const checkUserCoresAccess = ({
  user,
  requiredRole,
}: {
  user: Pick<User, 'id' | 'coresRole'>;
  requiredRole: CoresRole;
}): boolean => {
  if (!user) {
    return false;
  }

  if (typeof user.coresRole !== 'number') {
    return false;
  }

  if (user.coresRole < requiredRole) {
    return false;
  }

  return true;
};

export const checkCoresAccess = async ({
  ctx,
  userId,
  requiredRole,
}: {
  ctx: Context;
  userId: string;
  requiredRole: CoresRole;
}): Promise<boolean> => {
  const user = await ctx.dataLoader.user.load({
    userId,
    select: ['id', 'coresRole'],
  });

  if (!user) {
    return false;
  }

  return checkUserCoresAccess({ user, requiredRole });
};
