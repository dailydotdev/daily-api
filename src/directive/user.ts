import { DataSource } from 'typeorm';
import { FastifyLoggerInstance } from 'fastify';
import {
  Alerts,
  ArticlePost,
  Bookmark,
  BookmarkList,
  Comment,
  DevCard,
  Feed,
  Post,
  PostReport,
  Settings,
  SourceDisplay,
  SourceRequest,
  User,
  View,
} from '../entity';

export const deleteUser = async (
  con: DataSource,
  logger: FastifyLoggerInstance,
  userId: string,
  messageId?: string,
) => {
  try {
    await con.transaction(async (entityManager): Promise<void> => {
      await entityManager.getRepository(View).delete({ userId });
      await entityManager.getRepository(Alerts).delete({ userId });
      await entityManager.getRepository(BookmarkList).delete({ userId });
      await entityManager.getRepository(Bookmark).delete({ userId });
      await entityManager.getRepository(Comment).update(
        { userId },
        {
          userId: '404',
        },
      );
      await entityManager.getRepository(Comment).delete({ userId });
      await entityManager.getRepository(DevCard).delete({ userId });
      await entityManager.getRepository(Feed).delete({ userId });
      await entityManager.getRepository(PostReport).delete({ userId });
      await entityManager.getRepository(Settings).delete({ userId });
      await entityManager.getRepository(SourceDisplay).delete({ userId });
      await entityManager.getRepository(SourceRequest).delete({ userId });
      await entityManager
        .getRepository(ArticlePost)
        .update({ authorId: userId }, { authorId: null });
      // Manually set shared post to 404 dummy user
      await entityManager
        .getRepository(Post)
        .update({ authorId: userId }, { authorId: '404' });
      await entityManager
        .getRepository(Post)
        .update({ scoutId: userId }, { scoutId: null });
      await entityManager.getRepository(User).delete(userId);
    });
    if (logger) {
      logger.info(
        {
          userId,
          messageId,
        },
        'deleted user',
      );
    }
  } catch (err) {
    if (logger) {
      logger.error(
        {
          userId,
          messageId,
          err,
        },
        'failed to delete user',
      );
    }
    throw err;
  }
};
