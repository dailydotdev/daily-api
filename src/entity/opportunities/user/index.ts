export * from './OpportunityUser';
export * from './OpportunityUserRecruiter';
