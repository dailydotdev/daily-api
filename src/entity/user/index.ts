export * from './User';
export * from './UserPost';
export * from './UserState';
export * from './UserAction';
export * from './UserStreak';
export * from './UserPersonalizedDigest';
export * from './UserMarketingCta';
