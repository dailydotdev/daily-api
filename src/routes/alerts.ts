import { FastifyInstance } from 'fastify';
import { injectGraphql } from '../compatibility/utils';

export default async function (fastify: FastifyInstance): Promise<void> {
  fastify.get('/', async (req, res) => {
    const query = `{
      userAlerts {
        filter
        rankLastSeen
        myFeed
        companionHelper
        lastChangelog
        lastBanner
        squadTour
        showGenericReferral
        showStreakMilestone
        lastBootPopup
        lastFeedSettingsFeedback
      }
    }`;

    return injectGraphql(
      fastify,
      { query },
      (obj) => obj['data']['userAlerts'],
      req,
      res,
    );
  });
}
