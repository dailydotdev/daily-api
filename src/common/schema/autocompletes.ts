import z from 'zod';
import { normalizeContentForDeduplication } from '../../entity/posts/hooks';
import { AutocompleteType } from '../../entity';

export const autocompleteBaseSchema = z.object({
  query: z.string().transform(normalizeContentForDeduplication),
  limit: z.number().min(1).max(50).default(20),
});

export const autocompleteSchema = autocompleteBaseSchema.extend({
  type: z.enum(AutocompleteType),
});
