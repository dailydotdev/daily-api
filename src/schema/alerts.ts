import { Alerts, ALERTS_DEFAULT, UserActionType } from '../entity';
import { IResolvers } from '@graphql-tools/utils';
import { traceResolvers } from './trace';
import { Context } from '../Context';
import { DataSource } from 'typeorm';
import { insertOrIgnoreAction } from './actions';
import { GQLEmptyResponse } from './common';

interface GQLAlerts {
  filter: boolean;
}

interface GQLUpdateAlertsInput extends Partial<GQLAlerts> {
  filter?: boolean;
  myFeed?: string;
  lastBootPopup?: Date;
}

export const typeDefs = /* GraphQL */ `
  """
  Alerts to display to the relevant user
  """
  type Alerts {
    """
    Status to display for filter red dot
    """
    filter: Boolean!

    """
    Date last seen of the rank achievement popup
    """
    rankLastSeen: DateTime

    """
    Wether to show the my feed alert (migrated/created/null)
    migrated: The user has existing filters so we created myFeed for them
    created: The user applied filters himself
    null: The user clicked to not show the alert anymore
    """
    myFeed: String

    """
    Wether to show the companion helper
    Default = true, meaning yes show it
    Once the user has seen it once, we set this value to false
    """
    companionHelper: Boolean!

    """
    Date of the last changelog user saw
    """
    lastChangelog: DateTime

    """
    Date of the last banner user saw
    """
    lastBanner: DateTime

    """
    Whether to show the squad tour and sync across devices
    """
    squadTour: Boolean!

    """
    Whether to show the generic referral reminder
    """
    showGenericReferral: Boolean!

    """
    Whether to show the reading streaks milestone banner
    """
    showStreakMilestone: Boolean!

    """
    Date of the last time user saw a boot popup
    """
    lastBootPopup: DateTime

    """
    Date of the last time user dismissed or answered the feed survey
    """
    lastFeedSettingsFeedback: DateTime
  }

  input UpdateAlertsInput {
    """
    Status to display for filter red dot
    """
    filter: Boolean

    """
    Date last seen of the rank achievement popup
    """
    rankLastSeen: DateTime

    """
    Status for the My Feed alert
    """
    myFeed: String

    """
    Status to display for companion helper
    """
    companionHelper: Boolean

    """
    Date of the last changelog user saw
    """
    lastChangelog: DateTime

    """
    Date of the last banner user saw
    """
    lastBanner: DateTime

    """
    Whether to show the squad tour and sync across devices
    """
    squadTour: Boolean

    """
    Whether to show the reading streaks milestone banner
    """
    showStreakMilestone: Boolean

    """
    Date of the last time user saw a boot popup
    """
    lastBootPopup: DateTime
  }

  extend type Mutation {
    """
    Update the alerts for user
    """
    updateUserAlerts(data: UpdateAlertsInput!): Alerts! @auth

    """
    Update the last referral reminder
    """
    updateLastReferralReminder: EmptyResponse! @auth

    """
    Update the boot popup
    """
    updateLastBootPopup: EmptyResponse! @auth
    """
    Reset feed settings feedback (survey) reminder
    """
    updateFeedFeedbackReminder: EmptyResponse @auth
  }

  extend type Query {
    """
    Get the alerts for user
    """
    userAlerts: Alerts!
  }
`;

/**
 * Remove the flags from the alerts object
 * @param alerts
 */
export const saveReturnAlerts = (alerts: Alerts) => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { flags, ...data } = alerts;
  return data;
};

interface GQLAlertInputInternalInput {
  showGenericReferral?: boolean;
  lastFeedSettingsFeedback?: Date;
}

export const updateAlerts = async (
  con: DataSource,
  userId: string,
  data: GQLUpdateAlertsInput & GQLAlertInputInternalInput,
  flags?: Alerts['flags'],
): Promise<GQLAlerts> => {
  const repo = con.getRepository(Alerts);
  const alerts = await repo.findOneBy({ userId });

  if (data.filter === false) {
    await insertOrIgnoreAction(con, userId, UserActionType.MyFeed);
  }

  if (!alerts) {
    return saveReturnAlerts(
      await repo.save({
        userId,
        ...data,
        flags: flags ? flags : undefined,
      }),
    );
  }

  return saveReturnAlerts(
    await repo.save({
      ...alerts,
      ...data,
      flags: flags ? { ...alerts.flags, ...flags } : undefined,
    }),
  );
};

export const getAlerts = async (
  con: DataSource,
  userId: string,
): Promise<Omit<Alerts, 'flags'>> => {
  const alerts = await con.getRepository(Alerts).findOneBy({
    userId,
  });
  if (alerts) {
    return saveReturnAlerts(alerts);
  }
  return ALERTS_DEFAULT as Alerts;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const resolvers: IResolvers<any, Context> = traceResolvers({
  Mutation: {
    updateUserAlerts: async (
      _,
      { data }: { data: GQLUpdateAlertsInput },
      ctx,
    ): Promise<GQLAlerts> => updateAlerts(ctx.con, ctx.userId, data),
    updateLastReferralReminder: async (
      _,
      __,
      ctx,
    ): Promise<GQLEmptyResponse> => {
      await updateAlerts(
        ctx.con,
        ctx.userId,
        {
          showGenericReferral: false,
        },
        { lastReferralReminder: new Date() },
      );
      return { _: true };
    },
    updateLastBootPopup: async (_, __, ctx): Promise<GQLEmptyResponse> => {
      await updateAlerts(ctx.con, ctx.userId, { lastBootPopup: new Date() });
      return { _: true };
    },
    updateFeedFeedbackReminder: async (
      _,
      __,
      ctx,
    ): Promise<GQLEmptyResponse> => {
      await updateAlerts(ctx.con, ctx.userId, {
        lastFeedSettingsFeedback: new Date(),
      });
      return { _: null };
    },
  },
  Query: {
    userAlerts: (_, __, ctx): Promise<GQLAlerts> | GQLAlerts =>
      getAlerts(ctx.con, ctx.userId),
  },
});
