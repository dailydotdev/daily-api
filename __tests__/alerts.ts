import { FastifyInstance } from 'fastify';
import {
  Alerts,
  ALERTS_DEFAULT,
  User,
  UserAction,
  UserActionType,
} from '../src/entity';
import request from 'supertest';
import {
  authorizeRequest,
  disposeGraphQLTesting,
  GraphQLTestClient,
  GraphQLTestingState,
  initializeGraphQLTesting,
  MockContext,
  saveFixtures,
  testMutationErrorCode,
} from './helpers';
import createOrGetConnection from '../src/db';
import { DataSource } from 'typeorm';
import { saveReturnAlerts } from '../src/schema/alerts';
import { usersFixture } from './fixture/user';
import { isSameDay, subDays } from 'date-fns';

let app: FastifyInstance;
let con: DataSource;
let state: GraphQLTestingState;
let client: GraphQLTestClient;
let loggedUser: string = null;

beforeAll(async () => {
  con = await createOrGetConnection();
  state = await initializeGraphQLTesting(
    () => new MockContext(con, loggedUser),
  );
  client = state.client;
  app = state.app;
});

afterAll(() => disposeGraphQLTesting(state));

beforeEach(async () => {
  loggedUser = null;
  await saveFixtures(con, User, usersFixture);
});

describe('query userAlerts', () => {
  const QUERY = `{
    userAlerts {
      filter
      rankLastSeen
      myFeed
      companionHelper
      lastChangelog
      lastBanner
      squadTour
      showGenericReferral
      showStreakMilestone
      lastBootPopup
      lastFeedSettingsFeedback
    }
  }`;

  it('should return alerts default values if anonymous', async () => {
    const res = await client.query(QUERY);
    res.data.userAlerts.changelog = false;
    res.data.userAlerts.banner = false;
    res.data.userAlerts.bootPopup = false;
    expect(res.data.userAlerts).toEqual({
      ...ALERTS_DEFAULT,
      lastBanner: res.data.userAlerts.lastBanner,
      lastChangelog: res.data.userAlerts.lastChangelog,
      lastFeedSettingsFeedback: res.data.userAlerts.lastFeedSettingsFeedback,
    });
  });

  it('should return user alerts', async () => {
    loggedUser = '1';

    const repo = con.getRepository(Alerts);
    const alerts = repo.create({
      userId: '1',
      filter: true,
      flags: { lastReferralReminder: new Date('2023-02-05 12:00:00') },
    });
    await repo.save(alerts);
    const expected = saveReturnAlerts(
      await repo.findOneByOrFail({ userId: '1' })!,
    );
    const res = await client.query(QUERY);

    delete expected.userId;

    expect(res.data.userAlerts).toEqual({
      ...expected,
      lastBanner: expected.lastBanner.toISOString(),
      lastChangelog: expected.lastChangelog.toISOString(),
      lastFeedSettingsFeedback: expected.lastFeedSettingsFeedback.toISOString(),
    });
  });
});

describe('mutation updateUserAlerts', () => {
  const MUTATION = (extra = '') => `
    mutation UpdateUserAlerts($data: UpdateAlertsInput!) {
      updateUserAlerts(data: $data) {
        filter
        rankLastSeen
        myFeed
        companionHelper
        squadTour
        ${extra}
      }
    }
  `;

  it('should not authorize when not logged in', () =>
    testMutationErrorCode(
      client,
      {
        mutation: MUTATION(),
        variables: { data: { filter: false } },
      },
      'UNAUTHENTICATED',
    ));

  it('should create user alerts when does not exist', async () => {
    loggedUser = '1';
    const res = await client.mutate(MUTATION(), {
      variables: { data: { filter: false } },
    });
    expect(res.data).toMatchSnapshot();
  });

  it('should create user action type for my feed if alert is false', async () => {
    loggedUser = '1';
    const res = await client.mutate(MUTATION(), {
      variables: { data: { filter: false } },
    });
    const completed = await con
      .getRepository(UserAction)
      .findOneBy({ userId: '1', type: UserActionType.MyFeed });
    expect(completed).toBeTruthy();
    expect(res.data).toMatchSnapshot();
  });

  it('should update alerts of user', async () => {
    loggedUser = '1';

    const rankLastSeenOld = new Date('2020-09-21T07:15:51.247Z');
    const repo = con.getRepository(Alerts);
    await repo.save(
      repo.create({
        userId: '1',
        filter: true,
        rankLastSeen: rankLastSeenOld,
        myFeed: 'created',
        companionHelper: true,
        squadTour: true,
      }),
    );

    const rankLastSeen = new Date('2020-09-22T12:15:51.247Z');
    const res = await client.mutate(MUTATION(), {
      variables: {
        data: {
          rankLastSeen: rankLastSeen.toISOString(),
          myFeed: 'created',
          companionHelper: false,
          squadTour: false,
        },
      },
    });
    const completed = await con
      .getRepository(UserAction)
      .findOneBy({ userId: '1', type: UserActionType.MyFeed });

    expect(completed).toBeFalsy();
    expect(res.data).toMatchSnapshot();
  });

  it('should not update showGenericReferral alerts of user via updateAlerts', async () => {
    loggedUser = '1';

    const res = await client.mutate(MUTATION('showGenericReferral'), {
      variables: { data: { showGenericReferral: false } },
    });
    expect(res.errors).toBeTruthy();
    expect(res.errors[0].message).toEqual('Unexpected error');
  });
});

describe('dedicated api routes', () => {
  describe('GET /alerts', () => {
    it('should return user alerts', async () => {
      const repo = con.getRepository(Alerts);
      const alerts = repo.create({
        userId: '1',
        myFeed: 'created',
      });
      await repo.save(alerts);
      const expected = saveReturnAlerts(
        await repo.findOneByOrFail({ userId: '1' })!,
      );
      delete expected['userId'];

      loggedUser = '1';
      const res = await authorizeRequest(
        request(app.server).get('/alerts'),
      ).expect(200);
      expect(res.body).toEqual({
        ...expected,
        lastBanner: expected['lastBanner'].toISOString(),
        lastChangelog: expected['lastChangelog'].toISOString(),
        lastFeedSettingsFeedback:
          expected['lastFeedSettingsFeedback'].toISOString(),
      });
    });
  });
});

describe('mutation updateLastReferralReminder', () => {
  const MUTATION = `
    mutation UpdateLastReferralReminder {
      updateLastReferralReminder {
        _
      }
    }
  `;

  it('should not authorize when not logged in', () =>
    testMutationErrorCode(
      client,
      {
        mutation: MUTATION,
      },
      'UNAUTHENTICATED',
    ));

  it('should update the last referral reminder and flags', async () => {
    loggedUser = '1';
    const date = new Date();
    const res = await client.mutate(MUTATION);
    expect(res.errors).toBeFalsy();
    const alerts = await con.getRepository(Alerts).findOneBy({ userId: '1' });
    expect(alerts.showGenericReferral).toEqual(false);
    expect(alerts.flags.lastReferralReminder).not.toBeNull();
    expect(
      new Date(alerts.flags.lastReferralReminder).getTime(),
    ).toBeGreaterThanOrEqual(+date);
  });

  it('should update the last referral reminder and flags but keep existing flags', async () => {
    loggedUser = '1';

    // eslint-disable-next-line
    // @ts-ignore
    await con.getRepository(Alerts).save({
      userId: loggedUser,
      flags: { existingFlag: 'value1' },
    });

    const res = await client.mutate(MUTATION);
    expect(res.errors).toBeFalsy();
    const alerts = await con.getRepository(Alerts).findOneBy({ userId: '1' });
    expect(alerts.showGenericReferral).toEqual(false);
    expect(alerts.flags).toEqual({
      existingFlag: 'value1',
      lastReferralReminder: alerts.flags.lastReferralReminder,
    });
  });
});

describe('updateFeedFeedbackReminder', () => {
  const MUTATION = `
    mutation UpdateFeedFeedbackReminder {
      updateFeedFeedbackReminder {
        _
      }
    }
  `;

  it('should not authorize when not logged in', () =>
    testMutationErrorCode(client, { mutation: MUTATION }, 'UNAUTHENTICATED'));

  it('should reset the feed settings feedback reminder', async () => {
    loggedUser = '1';

    await con
      .getRepository(Alerts)
      .update(
        { userId: '1' },
        { lastFeedSettingsFeedback: subDays(new Date(), 1) },
      );

    const res = await client.mutate(MUTATION);
    expect(res.errors).toBeFalsy();
    const alerts = await con.getRepository(Alerts).findOneBy({ userId: '1' });
    expect(alerts.lastFeedSettingsFeedback).toBeTruthy();
    expect(isSameDay(alerts.lastFeedSettingsFeedback, new Date())).toBeTruthy();
  });
});
