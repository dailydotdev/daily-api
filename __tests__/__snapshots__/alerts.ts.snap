// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutation updateUserAlerts should create user action type for my feed if alert is false 1`] = `
Object {
  "updateUserAlerts": Object {
    "companionHelper": true,
    "filter": false,
    "myFeed": null,
    "rankLastSeen": null,
    "squadTour": true,
  },
}
`;

exports[`mutation updateUserAlerts should create user alerts when does not exist 1`] = `
Object {
  "updateUserAlerts": Object {
    "companionHelper": true,
    "filter": false,
    "myFeed": null,
    "rankLastSeen": null,
    "squadTour": true,
  },
}
`;

exports[`mutation updateUserAlerts should update alerts of user 1`] = `
Object {
  "updateUserAlerts": Object {
    "companionHelper": false,
    "filter": true,
    "myFeed": "created",
    "rankLastSeen": "2020-09-22T12:15:51.247Z",
    "squadTour": false,
  },
}
`;
